server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java

```diff
@@ -798,4 +752,202 @@ public void testUuidCreationLogging() {
             )
         );
     }
+
+    // =============== Shard Blobs to Delete Tests =================
+
+    /*
+        There is sufficient heap space to perform all operations on shardBlobsToDelete
+     */
+    @TestLogging(reason = "test includes assertions about logging", value = "org.elasticsearch.repositories.blobstore:WARN")
+    public void testShardBlobsToDeleteWithSufficientHeapSpace() {
+        // When the heap memory is above 2GB (Integer.MAX_VALUE) then we expect it to be limited to 2GB
+        long heapMemory = randomLongBetween(Integer.MAX_VALUE / 2, Integer.MAX_VALUE * 2L);
+        Settings.Builder settings = Settings.builder().put("repositories.blobstore.max_shard_delete_results_size", heapMemory + "b");
+        final var repo = setupRepo(settings);
+        try (var shardBlobsToDelete = repo.new ShardBlobsToDelete(settings.build())) {
+            // Ensure the logging is as expected
+            try (var mockLog = MockLog.capture(BlobStoreRepository.class)) {
+                // We expect every write to succeed
+                addFailureToCleanUpDanglingBlobsUnseenExpectation(mockLog);
+
+                final var expectedShardGenerations = ShardGenerations.builder();
+                final var expectedBlobsToDelete = new HashSet<String>();
+
+                final var countDownLatch = new CountDownLatch(1);
+                int blobCount = 0;
+                try (var refs = new RefCountingRunnable(countDownLatch::countDown)) {
+                    for (int index = between(0, 1000); index > 0; index--) {
+                        final var indexId = new IndexId(randomIdentifier(), randomUUID());
+                        for (int shard = between(1, 30); shard > 0; shard--) {
+                            final var shardId = shard;
+                            final var shardGeneration = new ShardGeneration(randomUUID());
+                            expectedShardGenerations.put(indexId, shard, shardGeneration);
+                            final var blobsToDelete = generateRandomBlobsToDelete(0, 100);
```

Comment:
Max 1000 indices × max 30 shards × max 100 blobs is a max of 3M items. That seems like a lot. Does this really give us much more coverage than a smaller test?

A much more interesting test here would be to see what happens if we stop writing just shy of the limit, such that the final flush pushes us over. Could we instead pick a lower limit, write until we get very close to the limit (according to `shardBlobsToDelete.sizeInBytes()`) and then verify that we didn't lose anything?